src/services/exit_order.py

```diff
@@ -1,24 +1,285 @@
-from src.web3py.typings import Web3
+from collections import defaultdict
+from dataclasses import dataclass
+
+from eth_typing import Address
+
+from src.modules.submodules.typings import ChainConfig
 
 from src.typings import BlockStamp
+from src.web3py.extentions.lido_validators import LidoValidator, NodeOperator, NodeOperatorIndex
+from src.web3py.typings import Web3
+
+FAR_FUTURE_EPOCH = 2 ** 64 - 1
+VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS = 3600  # todo: get from contract
+
+
+@dataclass
+class NodeOperatorPredictableState:
+    predictable_validators_total_age: int
+    predictable_validators_count: int
+    targeted_validators: int
+    delayed_validators: int
 
 
 class ValidatorsExit:
     """
-    Exit order is:
-    1. if NO active_keys < NO current keys
-    2. Stacking > 1% remove by sum validators age
-    3. By active validators numbers
+    Exit order predicates sequence:
+    1. Validator whose operator with the lowest number of delayed validators
+    2. Validator whose operator with the highest number of targeted validators
+    3. Validator whose operator with the highest stake weight
+    4. Validator whose operator with the highest number of validators
+    5. Validator with the lowest activation epoch
+    6. Validator with the lowest index
     """
-    def __init__(self, blockstamp: BlockStamp, w3: Web3,  max_size: int):
+
+    def __init__(
+        self,
+        blockstamp: BlockStamp,
+        w3: Web3,
+        c_conf: ChainConfig,
+        exitable_lido_validators: list[LidoValidator],
+        max_validators_to_exit: int,
+    ):
         self.blockstamp = blockstamp
         self.w3 = w3
+        self.c_conf = c_conf
+
+        self.exitable_lido_validators = exitable_lido_validators
+        self.max_validators_to_exit = max_validators_to_exit
 
-        # Iterator should contain only next validators
-        self.max_size = max_size
+        ###########
+        # todo: remove after keys-api fix. Key should contain staking module id, not only address
+        operators = self.w3.lido_validators.get_lido_node_operators(blockstamp)
+        self.staking_module_id = {
```

Comment:
Add class attrs with typing for this and the next attributes

Reply:
fixed